### PR TITLE
ephemeral pvc-backed cow disk image support

### PR DESCRIPTION
**Pull request by mpolednik**
This PR adds a new API for so called transient disks. Transient disk is a disk that overlays regular disk image. We can write to such disk without having the changes propagated to the base image. The API for that is a new VolumeSource called TransientVolumeSource that nests supported VolumeSources (indirectly, so only 1 level of transient overlays is supported).

The initial version aims to keep the COW image on the node's disk (not memory), created through separate container and shared with the VM container via path. The image size is not arbitrarily limited, but it does respect any limits imposed by Kubernetes EmptyDir (these are, by default, nil -- meaning that the image can grow arbitrarily large, up to available disk space). Images provided elsewhere (through PVs, separate network volumes etc.) are not in scope of this PR.

API example -- given regular PVC-backed storage:

```yaml
  volumes:
    - name: pvcvolume
      persistentVolumeClaim:
        claimName: disk-alpine
```

we are able to overlay that by nesting persistentVolumeClaim under `transient` key:

```yaml
  volumes:
    - name: pvcvolume
      transient:
        persistentVolumeClaim:
          claimName: disk-alpine
```

The main driver for this feature is CI, where we have to avoid "tainting" the base images while still allowing writes to the overlay image.

### CODE REVIEW

**Comment by mpolednik** [comment]
TODO: see if the qemu-img is needed; it most likely is not and could be replaced by BackingStore within libvirt

**Comment by fabiand** [comment]
Please provide a good description of the change in the PR description / initial comment.

File: pkg/api/v1/schema.go [hunk 1]
```diff
@@ -211,6 +211,9 @@ type VolumeSource struct {
 	// More info: https://kubevirt.gitbooks.io/user-guide/registry-disk.html
 	// +optional
 	RegistryDisk *RegistryDiskSource `json:"registryDisk,omitempty"`
+	// Transient is a special volume source that "wraps" specified source and provides copy-on-write image on top of it.
+	// +optional
+	Transient *VolumeSource `json:"transient,omitempty"`
```
[the remark this answers is not in the document]
Comment (rmohr) [reply pkg/api/v1/schema.go]: I would want to keep the nested approach, but I would have an extra `VolumeSource` struct which only contains disk sources which support transient disks (pvc right now).

Also we want the backing store feature from libvirt, no need to expect qcows, libvirt can set up everything for us and it should work for different source types (file, network, ...). Like discussed offline.

[the remark this answers is not in the document]
Comment (mpolednik) [reply pkg/api/v1/schema.go]: The last offline "discussion" was centered around the fact that libvirt doesn't set up anything for us wrt backingStore. Libvirt just complains that the image isn't there.

[the remark this answers is not in the document]
Comment (rmohr) [reply pkg/api/v1/schema.go]: too bad. Then the pr goes into the right direction.

[the remark this answers is not in the document]
Comment (rmohr) [reply pkg/api/v1/schema.go]: still we need the extra struct which only allow pvcs.

[the remark this answers is not in the document]
Comment (fabiand) [reply pkg/api/v1/schema.go]: I think it applies to all PV (raw or block) mounted through kubelet.

[the remark this answers is not in the document]
Comment (mpolednik) [reply pkg/api/v1/schema.go]: That makes this PR depend on #734.

**Comment by simon3z** [comment]
> Please provide a good description of the change in the PR description / initial comment.

@mpolednik I think it would be important to know more on the motivations for supporting transient disks (it seems tricky so we should try to understand the story)

**Comment by mpolednik** [comment]
@simon3z Done.

File: pkg/transient-disk/transient-disk.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+				},
+			})
+		}
+	}
+
+	return containers, nil
+}
+
+func getFilePathByVolume(vm *v1.VirtualMachine, src v1.Volume) (string, string, error) {
+	if src.RegistryDisk == nil {
```
Comment (davidvossel) [inline pkg/transient-disk/transient-disk.go]: this line looks like it prevents transient disks from working. I think you want to change this to look for the PVC file now.

File: pkg/virt-controller/services/template.go [hunk 3]
```diff
@@ -158,6 +175,26 @@ func (t *templateService) RenderLaunchManifest(vm *v1.VirtualMachine) (*kubev1.P
 			},
 		},
 	})
+
+	transientVolumes, err := transientdisk.GenerateVolumes(vm)
```
Comment (davidvossel) [inline pkg/virt-controller/services/template.go]: all this generate container/volume logic needs a test case in pkg/virt-controller/services/template.go

File: pkg/transient-disk/transient-disk.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+	return fmt.Sprintf("%s", mountBaseDir)
+}
+func generateVolumeMountDir(vm *v1.VirtualMachine) string {
+	baseDir := generateVMBaseDir(vm)
+	return fmt.Sprintf("%s", baseDir)
+}
+
+func GetFilePath(vm *v1.VirtualMachine, volumeName string) (string, string, error) {
+	volumeMountDir := generateVolumeMountDir(vm)
+	return volumeMountDir + "/" + volumeName + ".qcow2", "qcow2", nil
```
Comment (davidvossel) [inline pkg/transient-disk/transient-disk.go]: there's no reason to return an error here if 'nil' is the only possibility for error.

**Comment by davidvossel** [comment]
@mpolednik I began questioning if our use of the term "transient" here is accurate. 

While 'transient' does mean 'short lived',  I've typically heard the term used to describe something that is intermittent... like a 'transient' failure that occasionally appears and is self corrected.

Example: *"During a transient failure the database reads are eventually consistent."*

With respects to storage, the two terms that appear most common are **persistent** and **ephemeral**.  aws and openstack use the term ephemeral when describing storage that does not outlive the lifecycle of a running VM instance. 

Example: *"We use ephemeral EBS volumes on our EC2 instances because our applications are stateless"*

File: pkg/transient-disk/transient-disk.go (hunk 4 above)
Comment (mpolednik) [reply pkg/transient-disk/transient-disk.go]: This function actually tries to imitate GetFilePath from registrydisk (they could be behind interface).

**Comment by mpolednik** [comment]
@davidvossel That makes for an interesting conflict with what we already call ephemeral volumes. I don't have hard feeling about the name itself, renaming currentEpehemeral -> persistent and transient -> ephemeral is fine by me (and could be done as a part of this PR).

**Comment by simon3z** [comment]
> The main driver for this feature is CI, where we have to avoid "tainting" the base images while still allowing writes to the overlay image.

Thanks @mpolednik!

Can you add few words on where we plan to keep the cow layer (e.g. tmpfs, same PV, additional PV, host disks, etc.) and the considerations/limitations based on that? (e.g. there may be limits on how large the cow layer could grow, etc.)

**Comment by mpolednik** [comment]
@simon3z added a paragraph highlighting these details. It's worth noting that since current scope of the PR is CI, we haven't thought of imposing any limitation on the max size or other disk attributes. @fabiand maybe the limits are worth a consideration?

**Comment by simon3z** [comment]
In general I am a little skeptical to add an entirely new feature just for CI for two reasons:

1. it's hard to mark it as just CI feature, someone will use it (at their own risk) and will bump in limitations creating some noise (issues, questions, etc.)
2. CI is going to test a case that we don't support instead of testing what we support, the result is that the feature created for CI (with all it usability limitation) ironically could result more tested/stable than the officially supported ones

In the specifics I think that the `emptyDir` approach is a little bit tricky because it depends on configurations that could vary very much in deployments and solutions (e.g. OpenShift has its own defaults).

Anyway I am not familiar with what specific CI problem is trying to solve (I can just try to guess) so I may not be able to understand the compromises that we're willing to make here.

**Comment by rmohr** [comment]
> In general I am a little skeptical to add an entirely new feature just for CI for two reasons:

CI and testing are important (and often the first) users of a project

> it's hard to mark it as just CI feature, someone will use it (at their own risk) and will bump in limitations creating some noise (issues, questions, etc.)

If it is also useful outside of the CI context, then it is valuable in my opinion.

> CI is going to test a case that we don't support instead of testing what we support, the result is that the feature created for CI (with all its usability limitation) ironically could result more tested/stable than the officially supported ones

CI needs it to test windows installations. I don't want to modify 30 gig images and I don't want to rely on snapshot/clone support of storage at this testing stage (windows support is independent of storage support/features and I don't know when we get what).

Apart from that it is in general useful. I also have a use-cases associated with the cluster auto-scaler. There are two typical use-cases on how one uses cloud images:

 * one boots a read-only image and it is expected that another disk is mounted (addressed by #694).
 * one boots a writable image and expects that the base-image will not be modified (addressed here).

An HA-storage solution with its own heavy requirements works too. This provides a features which allows addressing some of these use-cases in a very lightweight fashion which can be useful for a lot of people.

@simon3z I hope that clarifies that.

**Comment by davidvossel** [comment]
> That makes for an interesting conflict with what we already call ephemeral volumes. I don't have hard feeling about the name itself, renaming currentEpehemeral -> persistent and transient -> ephemeral is fine by me (and could be done as a part of this PR).

@mpolednik where do we already use the term ephemeral volumes in KubeVirt?

**Comment by rmohr** [comment]
> @mpolednik where do we already use the term ephemeral volumes in KubeVirt?

@davidvossel We use `ephemeral` in the user-guide in the context of registry disks. I think we are using it in a consistent way if we use it here too. So +1 for your desire to move to `ephemeral`.

**Comment by mpolednik** [comment]
Ack, I've confused ephemeral disks with ephemeral volumes.

**Comment by rmohr** [comment]
> @simon3z added a paragraph highlighting these details. It's worth noting that since current scope of the PR is CI, we haven't thought of imposing any limitation on the max size or other disk attributes. @fabiand maybe the limits are worth a consideration?

@mpolednik Starting without that is perfect for me in this context. I think we can take it into account in the `resource` section on the pod in the context of the `ephemeral-store` attribute there. Also #694 will need something like this.  We can tackle this in a common solution later.

**Comment by davidvossel** [comment]
> We use ephemeral in the user-guide in the context of registry disks. I think we are using it in a consistent way if we use it here too. So +1 for your desire to move to ephemeral.

Excellent. In the user guide, we'll just point out that registry disks are inherently ephemeral in nature, and that other volumes (like PVCs for now) are can be toggled to be ephemeral with writes backed by local storage if desired. I don't think that will be too confusing.

File: pkg/transient-disk/transient-disk.go [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+	return volumeMounts
+}
+
+func generateVolumeMountsForVolume(volume v1.Volume, vm *v1.VirtualMachine) []kubev1.VolumeMount {
+	var volumeMounts []kubev1.VolumeMount
+
+	volumeMounts = append(volumeMounts, kubev1.VolumeMount{
+		Name:      createBackingVolumeName(volume.Name),
+		MountPath: filepath.Join("/var/run/kubevirt-private", "vm-disks", volume.Name),
```
Comment (davidvossel) [inline pkg/transient-disk/transient-disk.go]: I'd like to see all this logic involved with building PVC file paths converged into a single location. We're building this path in template.go, converter.go, and here too now.

Comment (mpolednik) [reply pkg/transient-disk/transient-disk.go]: Oh yeah totally, I was thinking of pkg/pvc -- symmetric to pkg/transientdisk and pkg/registrydisk, where each "storage type" provides interface that the template uses. This case is the worst offender.

**Comment by fabiand** [comment]
@mpolednik yes - limits would be something that are worth to consider in future. we could combine it with local PVs which can be requested with resources - which mean sthat we also get available space guarantees.

**Comment by mpolednik** [comment]
Ran `make generate` against all commits; current travis error seems to be caused by

```
Get https://registry-1.docker.io/v2/library/fedora/manifests/27: Get https://auth.docker.io/token?scope=repository%3Alibrary%2Ffedora%3Apull&service=registry.docker.io: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
```

**Comment by rmohr** [comment]
@mpolednik restarted travis

File: pkg/ephemeral-disk/ephemeral-disk.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+	var containers []kubev1.Container
+
+	initialDelaySeconds := 2
+	timeoutSeconds := 5
+	periodSeconds := 5
+	successThreshold := 2
+	failureThreshold := 5
+
+	// Make VM Image Wrapper Containers
+	for _, volume := range vm.Spec.Volumes {
```
Comment (rmohr) [inline pkg/ephemeral-disk/ephemeral-disk.go]: Afaik all that is not necessary. kubevirt/libvirt already containes `qemu-img`. Just create an empty qcow2 file in the virt-launcher container and place it in the empty-dir directory for e.g. registry disks in a unique location. We would then follow the same pattern like the cloud-init volume (the whole disk content is predictable for virt-launcher).

**Comment by davidvossel** [comment]
> With directly calling qemu-img from within virt-launcher the code should get much simpler. It can work like the cloud-init code. No extra containers are needed for that.

@mpolednik I think Roman's right. we should be taking the cloud-init approach here. 

Take a look at the preStartHook() function in pkg/virt-launcher/virtwrap/manager.go. That should be the entry point for the ephemeral disk logic that generates the qcow2 backing file. You'd just need to loop through the ephemeral volumes in the VM spec there and create the backing files. All the template.go logic with volumes and containers could go away after that. 

Originally I liked the way you were using containers and volumes to generate the backing file up front before the VM start flow, but @rmohr pointed out that it's way more complicated and inefficient to take that approach. It also requires shipping another container with KubeVirt that we could avoid.

**Comment by davidvossel** [comment]
@mpolednik The new direction you're taking looks like what I was hoping to see.

File: tests/storage_test.go [hunk 7]
```diff
@@ -165,5 +165,29 @@ var _ = Describe("Storage", func() {
 				close(done)
 			}, 240)
 		})
+
+		Context("With ephemeral alpine PVC", func() {
```
Comment (cynepco3hahue) [inline tests/storage_test.go]: Move this under `Describe("Starting a VM...)` and change to `with`
I also miss test that 
- touch some file
- restart the VM with ephemeral PVC
- check that the file does not exist

Comment (mpolednik) [reply tests/storage_test.go]: The test literally is under Describe("Starting a VM block"), it's just different context. Agree about the test described above, that's on the list.

Comment (cynepco3hahue) [reply tests/storage_test.go]: @mpolednik srry, missed it

**Comment by mpolednik** [comment]
functests are now added and working (at least locally)

**Comment by rmohr** [comment]
ok to test

**Comment by rmohr** [comment]
@mpolednik travis failes because of #777, working on that.

File: tests/storage_test.go [hunk 8]
```diff
@@ -165,5 +170,78 @@ var _ = Describe("Storage", func() {
...
+				By("Checking that the VM console has expected output")
+				_, err = expecter.ExpectBatch([]expect.Batcher{
+					&expect.BExp{R: "Welcome to Alpine"},
+				}, 200*time.Second)
+				Expect(err).To(BeNil())
+
+				close(done)
+			}, 240)
+
+			It("should not persist data", func(done Done) {
```
Comment (rmohr) [inline tests/storage_test.go]: I fear you will have to be a little bit more creative here. Alpine and Cirros both boot read-only. That might actually be pretty tricky to solve without a storage provisioner. We could add an empy "PVC" which we clean after the tests. Not sure ...

File: pkg/ephemeral-disk/ephemeral-disk.go [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+	return nil
+}
+
+func GetFilePath(volumeName string) string {
+	volumeMountDir := generateVolumeMountDir(volumeName)
+	return filepath.Join(volumeMountDir, "disk.qcow2")
+}
+
+func CreateEphemeralImages(vm *v1.VirtualMachine) error {
```
Comment (rmohr) [inline pkg/ephemeral-disk/ephemeral-disk.go]: That function grieves for one or more unit-tests.

File: tests/storage_test.go (hunk 8 above)
Comment (mpolednik) [reply tests/storage_test.go]: They boot rw (tmpfs on / type tmpfs (rw,relatime,mode=755)), the tricky part is tmpfs...

Comment (rmohr) [reply tests/storage_test.go]: Oh that is great, then you can just mount the disk again on a different location and write there to it.

Comment (rmohr) [reply tests/storage_test.go]: oh that probably does not help, since if I bindmount it it will probably also point to the tmpfs.

Comment (mpolednik) [reply tests/storage_test.go]: And to make matters wrose, /dev/sda1 is formatted as iso9660. :/

File: pkg/virt-launcher/virtwrap/api/schema.go [hunk 10]
```diff
@@ -265,6 +266,16 @@ type DiskSourceHost struct {
 	Port string `xml:"port,attr,omitempty"`
 }
 
+type BackingStore struct {
```
Comment (rmohr) [inline pkg/virt-launcher/virtwrap/api/schema.go]: Can you also add a mapping test, so that we have the expected xml outcome tested?

Comment (mpolednik) [reply pkg/virt-launcher/virtwrap/api/schema.go]: On it as soon as I figure out the functional test (thanks for pointing that out, totally missed the tmpfs mount).

File: tests/storage_test.go [hunk 11]
```diff
@@ -165,5 +170,88 @@ var _ = Describe("Storage", func() {
...
+				expecter, _, err := tests.NewConsoleExpecter(virtClient, vm, "serial0", 10*time.Second)
+				Expect(err).ToNot(HaveOccurred())
+				defer expecter.Close()
+				_, err = expecter.ExpectBatch([]expect.Batcher{
+					&expect.BExp{R: "Welcome to Alpine"},
+					&expect.BSnd{S: "\n"},
+					&expect.BExp{R: "login"},
+					&expect.BSnd{S: "root\n"},
+					&expect.BExp{R: "#"},
+					// Because "/" is mounted on tmpfs, we need something that normally persists writes - /dev/sda2 is the EFI partition formatted as vFAT.
```
Comment (mpolednik) [inline tests/storage_test.go]: It's actually /dev/vda2 in functests.

**Comment by rmohr** [comment]
retest this please

**Comment by mpolednik** [comment]
@rmohr on it.

**Comment by mpolednik** [comment]
User guide update: https://github.com/kubevirt/user-guide/pull/29

**Pull request merged**